### PR TITLE
Protobuf support

### PR DESCRIPTION
**Pull request by s0l0ist**
## Description
Adding support to use protobufs instead of JSON objects/strings

## Affected Dependencies
This is a breaking change. The user will need to use the built-in mechanisms for serializing/deserializing the protobufs.

## How has this been tested?
- Run the bazel tests for all bindings

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

TODO:
- [x] C++
- [x] C
- [x] Go
- [x] Python
- [x] JavaScript

### CODE REVIEW

File: private_set_intersection/python/psi_bindings.cpp [hunk 1]
```diff
@@ -27,6 +28,28 @@ PYBIND11_MODULE(_psi_bindings, m) {
       "Filters";
 
   m.attr("__version__") = ::private_set_intersection::Package::kVersion;
+  py::class_<psi_proto::ServerSetup>(m, "PsiProtoServerSetup")
+      .def("clear_bits", &psi_proto::ServerSetup::clear_bits)
+      .def("bits", &psi_proto::ServerSetup::bits)
+      .def("set_allocated_bits", &psi_proto::ServerSetup::set_allocated_bits)
```
Comment (s0l0ist) [inline private_set_intersection/python/psi_bindings.cpp:34]: Should this be `set_bits` instead?

Comment (s0l0ist) [reply private_set_intersection/python/psi_bindings.cpp:34]: fixed.

File: WORKSPACE [hunk 2]
```diff
@@ -7,3 +7,5 @@ psi_preload()
 load("@org_openmined_psi//private_set_intersection:deps.bzl", "psi_deps")
 
 psi_deps()
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
```
[the remark this answers is not in the document]
Comment (s0l0ist) [reply WORKSPACE:11]: Ah you're right, it doesn't. Was a left over artifact from when I was experimenting with bazel + ts protos.

File: private_set_intersection/javascript/bindings/client.cpp [hunk 3]
```diff
@@ -34,56 +36,105 @@ EMSCRIPTEN_BINDINGS(PSI_Client) {
...
+            const std::size_t server_setup_length =
+                server_setup_array["length"].as<std::size_t>();
+            std::string server_setup_string(server_setup_length, '\0');
+            for (std::size_t i = 0; i < server_setup_length; i++) {
+              server_setup_string[i] = server_setup_array[i].as<std::uint8_t>();
+            }
+
+            const std::size_t server_response_length =
+                server_response_array["length"].as<std::size_t>();
+            std::string server_response_string(server_response_length, '\0');
```
Comment (schoppmp) [inline private_set_intersection/javascript/bindings/client.cpp:70]: If it was possible to get a pointer to the raw Uint8Array data from emscripten::val directly, we could use ParseFromArray below and avoid copying to a string. But I'm not sure if emscripten allows this.

File: private_set_intersection/javascript/bindings/utils.h [hunk 4]
```diff
@@ -34,6 +34,30 @@ emscripten::val ToJSObject(private_join_and_compute::StatusOr<T> statusor) {
...
 }
 
+template <typename T>
+emscripten::val ToSerializedJSObject(
+    private_join_and_compute::StatusOr<T> statusor) {
+  auto result = emscripten::val::object();
+  if (statusor.ok()) {
+    const T protobuf = statusor.ValueOrDie();
+    std::string byte_string;
+    protobuf.SerializeToString(&byte_string);
```
Comment (schoppmp) [inline private_set_intersection/javascript/bindings/utils.h:44]: You can avoid a copy by first allocating the std::vector<uint8_t> of the correct size and then using SerializeToArray to write to it directly.

File: private_set_intersection/javascript/src/__tests__/client.test.ts [hunk 5]
```diff
@@ -74,135 +76,322 @@ describe('PSI Client', () => {
...
+//   77, 212, 199,  54, 193,  72, 141,  23, 252, 194, 187,  27,
+//  150, 197, 124, 201, 183,   7, 100,  43, 197, 142,   6, 138,
+//   96,  73,  11,  38,  18,  33,   2, 101, 115, 165, 127, 201,
+//   74, 239,  33, 103,  94,  18,  43,  62,  50,  17, 223,  83,
+//  177,  25,  23, 196, 200, 192, 198, 115,  40,  84,  90, 137,
+//  228, 118,  16,  18,  33,   2,  82,  37, 166, 155,   4, 244,
+//  228, 122, 150, 169, 225,  71, 110, 213,  42,  18, 105, 146,
+//   58, 180, 149, 197,  47, 214,  81, 126, 113, 248,   5,  16,
+//   19, 194
+// ]))
```
Comment (schoppmp) [inline private_set_intersection/javascript/src/__tests__/client.test.ts:397]: Is this still needed?

File: .bazelrc [hunk 6]
```diff
@@ -1,5 +1,5 @@
-##### Native #####
- # No need to specify additional configs
+##### Global #####
+build --cxxopt=--std=c++17
```
[the remark this answers is not in the document]
Comment (s0l0ist) [reply .bazelrc:2]: Yes, for your reference this is needed for pybind11 binding syntax

File: private_set_intersection/javascript/bindings/client.cpp [hunk 7]
```diff
@@ -34,56 +36,105 @@ EMSCRIPTEN_BINDINGS(PSI_Client) {
                 optional_override([](const PsiClient& self,
                                      const emscripten::val& string_array) {
```
[the remark this answers is not in the document]
Comment (s0l0ist) [reply private_set_intersection/javascript/bindings/client.cpp:37]: Good catch, will update

File: private_set_intersection/javascript/bindings/client.cpp (hunk 3 above)
Comment (s0l0ist) [reply private_set_intersection/javascript/bindings/client.cpp:70]: I tried to use the array iterators provided by emscripten::val::array, but was not successful in getting it to work properly using `ParseFromArray`. 

There might a way to get a pointer to `val`, I would prefer to release what is working now and work on this optimization in the future.

File: private_set_intersection/cpp/psi_client.cpp [hunk 8]
```diff
@@ -110,56 +97,41 @@ StatusOr<std::vector<int64_t>> PsiClient::GetIntersection(
...
+                   BloomFilter::CreateFromProtobuf(server_setup));
 
-  // Decrypt all elements in the response.
-  if (!response.IsObject() || !response.HasMember("encrypted_elements")) {
-    return ::private_join_and_compute::InvalidArgumentError(
-        "`server_response` is malformed");
-  }
-  const auto response_array = response["encrypted_elements"].GetArray();
-  int64_t response_size = static_cast<int64_t>(response_array.Size());
+  const auto response_array = server_response.encrypted_elements();
```
[the remark this answers is not in the document]
Comment (s0l0ist) [reply private_set_intersection/cpp/psi_client.cpp:125]: the type is `google::protobuf::internal::RepeatedStringTypeTraits::RepeatedFieldType`, which I felt was too long and doesn't really make much sense as it is by itself.

File: private_set_intersection/javascript/bindings/utils.h (hunk 4 above)
Comment (s0l0ist) [reply private_set_intersection/javascript/bindings/utils.h:44]: I tried to implement your suggestion, but I don't think `SerializeToArray` doesn't accept a `vector` type.

File: private_set_intersection/cpp/psi_server.cpp [hunk 9]
```diff
@@ -78,84 +75,46 @@ StatusOr<std::string> PsiServer::CreateSetupMessage(
...
-  rapidjson::Value response_elements;
-  response_elements.SetArray();
-  for (int i = 0; i < num_client_elements; i++) {
-    response_elements.PushBack(
-        rapidjson::Value().SetString(reencrypted_elements[i].data(),
-                                     reencrypted_elements[i].size(),
-                                     response.GetAllocator()),
-        response.GetAllocator());
+    auto sorted_elements = response.encrypted_elements();
+    std::sort(sorted_elements.begin(), sorted_elements.end());
```
[the remark this answers is not in the document]
Comment (s0l0ist) [reply private_set_intersection/cpp/psi_server.cpp:115]: Yes, you're correct. We need to pre-sort then re-write back to the protobuf.

[the remark this answers is not in the document]
Comment (s0l0ist) [reply private_set_intersection/cpp/psi_server.cpp:117]: I've implemented a solution.

File: private_set_intersection/javascript/bindings/utils.h (hunk 4 above)
Comment (schoppmp) [reply private_set_intersection/javascript/bindings/utils.h:44]: You can get a raw pointer from a vector by calling vector.data().

File: private_set_intersection/cpp/psi_client.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (schoppmp) [reply private_set_intersection/cpp/psi_client.cpp]: You're right, that is a bit long. Then maybe just change the type to `const auto &` to make explicit that encrypted_elements() returns a reference.

File: private_set_intersection/javascript/bindings/client.cpp (hunk 3 above)
Comment (schoppmp) [reply private_set_intersection/javascript/bindings/client.cpp:69]: Sounds good to me.

**Pull request merged**